simple_history/manager.py

```diff
@@ -9,13 +9,114 @@
 )
 
 
+class HistoricalQuerySet(QuerySet):
+    """
+    Enables additional functionality when working with historical records.
+
+    For additional history on this topic, see:
+        - https://github.com/jazzband/django-simple-history/pull/229
+        - https://github.com/jazzband/django-simple-history/issues/354
+        - https://github.com/jazzband/django-simple-history/issues/397
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(HistoricalQuerySet, self).__init__(*args, **kwargs)
+        self._as_instances = False
+        self._pk_attr = self.model.instance_type._meta.pk.attname
+
+    def as_instances(self):
+        """
+        Return a queryset that generates instances instead of historical records.
+        Queries against the resulting queryset will translate `pk` into the
+        primary key field of the original type.
+
+        Returns a queryset.
+        """
+        if not self._as_instances:
+            result = self.exclude(history_type="-")
+            result._as_instances = True
+        else:
+            result = self._clone()
+        return result
+
+    def filter(self, *args, **kwargs):
+        """
+        If a `pk` filter arrives and the queryset is returning instances
+        then the caller actually wants to filter based on the original
+        type's primary key, and not the history_id (historical record's
+        primary key); this happens frequently with DRF.
+        """
+        if self._as_instances and "pk" in kwargs:
+            kwargs[self._pk_attr] = kwargs.pop("pk")
+        return super().filter(*args, **kwargs)
+
+    def latest_of_each(self):
+        """
+        Ensures results in the queryset are the latest historical record for each
+        primary key.  Deletions are not removed.
+
+        Returns a queryset.
+        """
+        # If using MySQL, need to get a list of IDs in memory and then use them for the
+        # second query.
+        # Does mean two loops through the DB to get the full set, but still a speed
+        # improvement.
```

Comment:
I didn't realize this wasn't new code. You can ignore the following if you'd like.


A speed improvement in what case(s)? If we don't have a repeatable test case to prove that, this comment isn't helpful.

Ideally, this should use the same code for all vendors. If there's a bug or performance problem, allowing the developer a hook to change how the query is built may be better.

Personally, I'd vote for the subquery approach.

Reply:
Most of the subquery code was simply moved into HistoricalQuerySet so it could be more selectively applied; and was not otherwise changed.